torchreid/apis/training.py

```diff
@@ -0,0 +1,106 @@
+"""
+ Copyright (c) 2021 Intel Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from copy import deepcopy
+
+import torchreid
+from torchreid.engine import build_engine
+from torchreid.optim import LrFinder
+from scripts.default_config import (lr_finder_run_kwargs,
+                                    lr_scheduler_kwargs, model_kwargs,
+                                    optimizer_kwargs, engine_run_kwargs,
+                                    engine_test_kwargs)
+from torchreid.utils import set_random_seed
+from scripts.script_utils import (build_datamanager, build_auxiliary_model,
+                                  put_main_model_on_the_device)
+
+
+def run_lr_finder(cfg, datamanager, model, optimizer, scheduler, classes,
+                  rebuild_model=True, gpu_num=1, split_models=False):
+    if rebuild_model:
+        tmp_model = model
+    else:
+        tmp_model = deepcopy(model)
+
+    engine = build_engine(cfg, datamanager, tmp_model, optimizer, scheduler, initial_lr=cfg.train.lr)
+    lr_finder = LrFinder(engine=engine, **lr_finder_run_kwargs(cfg))
+    aux_lr = lr_finder.process()
+
+    print(f"Estimated learning rate: {aux_lr}")
+    if cfg.lr_finder.stop_after:
+        print("Finding learning rate finished. Terminate the training process")
+        exit()
+
+    # reload all parts of the training
+    # we do not check classification parameters
+    # and do not get num_train_classes the second time
+    # since it's done above and lr finder cannot change parameters of the datasets
+    cfg.train.lr = aux_lr
+    cfg.lr_finder.enable = False
+    set_random_seed(cfg.train.seed, cfg.train.deterministic)
+    datamanager = build_datamanager(cfg, classes)
+    num_train_classes = datamanager.num_train_pids
+    if rebuild_model:
```

Comment:
why we need this parameter? Don't we need to rebuild the model always after the LR finder?

Reply:
In the task environment, the original model state should be restored instead of rebuilding the model from scratch.